### PR TITLE
Use one build definition to produce all Linux distro specific packages

### PR DESCRIPTION
**Pull request by wli3**
By using `./build.sh /t:AllLiuxDistrosNativeInstaller`

Make a sandbox folder after compile step and copy compiled artifact to it. Since the content will be Linux generic, use package step with Docker in different distros.
After finishing all the distro specific packaging, upload all of them at once

### CODE REVIEW

File: build/LinuxDistrosNativeInstaller.props [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<Project>
+  <ItemGroup>
+    <LinuxDistrosNeedNativeInstaller Include="ubuntu.16.04-x64">
+      <DockerFolder>ubuntu.16.04</DockerFolder>
+      <InstallerExtension>deb</InstallerExtension>
+      <RepositoryId>$(ubuntu1604x64RepositoryId)</RepositoryId>
+      <DistroNameInInstaller>ubuntu.16.04-x64</DistroNameInInstaller>
```
Comment (wli3) [inline build/LinuxDistrosNativeInstaller.props]: This field has duplicated logic, I don't know how to fix it. I cannot find a easy way for msbuild to look up a dictionary, or even better a function that map osname, rid, architect to this field.

**Comment by wli3** [comment]
@dotnet-bot test this

**Comment by wli3** [comment]
the perf is bad, 84.4 minutes. Compare with 1 dedicated lag will run 30 minutes. From the log, there should be some degree of parallelism. Need to keep investigating

**Comment by wli3** [comment]
@dotnet-bot Test this please

**Comment by wli3** [comment]
Copying more cache into sandbox, the perf improved to 57 minutes, close to slowest Windows lag

File: .gitignore [hunk 2]
```diff
@@ -97,6 +97,7 @@ dlldata.c
 project.lock.json
 artifacts/
 bin/
+artifactsSandBox/
```
Comment (johnbeisner) [inline .gitignore]: Can we call this something different? 
"artifacts" was removed - so "artifactsSandBox" doesn't seem desirable.

You move _everything_ into artifactsSandBox and then work from artifactsSandBox, correct?

How about "WorkingCopy" or something like that...

File: Directory.Build.props [hunk 3]
```diff
@@ -21,7 +21,9 @@ tools\TestAssetsDependencies\TestAssetsDependencies.csproj
     <NuGetPackagesDir>$(NUGET_PACKAGES)</NuGetPackagesDir>
     <NuGetPackagesDir Condition=" '$(NuGetPackagesDir)' == '' ">$(RepoRoot)/.nuget/packages</NuGetPackagesDir>
 
-    <CLIBuildDll>$(RepoRoot)/build_projects/dotnet-cli-build/bin/dotnet-cli-build.dll</CLIBuildDll>
+    <RelativeCLIBuildBinaries>build_projects/dotnet-cli-build/bin</RelativeCLIBuildBinaries>
+    <RelativeCLIBuilddllName>dotnet-cli-build.dll</RelativeCLIBuilddllName>
+    <CLIBuildDll>$(RepoRoot)/$(RelativeCLIBuildBinaries)/$(RelativeCLIBuilddllName)</CLIBuildDll>
```
Comment (johnbeisner) [inline Directory.Build.props]: CamelCase this: RelativeCLIBuildDllName

File: scripts/obtain/dotnet-install.sh [hunk 4]
```diff
@@ -656,7 +656,9 @@ install_dotnet() {
     say_verbose "Zip path: $zip_path"
 
     say "Downloading link: $download_link"
-    download "$download_link" $zip_path || download_failed=true
+    # Failures are expected in legacy case. Do not output to stderr, since
+    # the output stderr are considered an error
+    download "$download_link" $zip_path 2>&1 || download_failed=true
```
Comment (johnbeisner) [inline scripts/obtain/dotnet-install.sh]: Failures are expected in the _non-_legacy case

File: build/SandBoxAndPackageInParallel.proj [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  <Import Project="$(MSBuildThisFileDirectory)/LinuxDistrosNativeInstaller.props" />
+
+  <Target Name="RunInSandBoxAndPackage" >
+      <MSBuild Projects="$(MSBuildThisFileDirectory)/SandBoxAndPackage.proj" 
+      Targets="SandBoxAndPackage"
+            Properties="RepoRoot=$(RepoRoot);
+            LinuxDistrosNeedNativeInstaller=%(LinuxDistrosNeedNativeInstaller.Identity);
+            DockerFolder=%(LinuxDistrosNeedNativeInstaller.DockerFolder);
+            BuildConfiguration=$(BuildConfiguration);
+            InstallerOutputDirectory=$(InstallerOutputDirectory);
```
Comment (wli3) [inline build/SandBoxAndPackageInParallel.proj]: BuildInParallel='true'

Comment (wli3) [inline build/SandBoxAndPackageInParallel.proj]: BuildInParallel='true'

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (wli3) [reply .gitignore]: Changed to bin/WorkingCopy

Comment (wli3) [reply .gitignore]: Changed to bin/WorkingCopy

File: Directory.Build.props (hunk 3 above)
Comment (wli3) [reply Directory.Build.props]: Fixed

File: scripts/obtain/dotnet-install.sh (hunk 4 above)
Comment (wli3) [reply scripts/obtain/dotnet-install.sh]: Fixed

File: build/AllLinuxDistrosNativeInstaller.targets [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Layout" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildThisFileDirectory)/LinuxDistrosNativeInstaller.props" />
+
+  <Target Name="AllLiuxDistrosNativeInstaller" DependsOnTargets="Prepare;Compile;RunAllSandBoxAndPackage;MultiPublish">
```
[the remark this answers is not in the document]
Comment (wli3) [reply build/AllLinuxDistrosNativeInstaller.targets]: Fixed

File: build/AllLinuxDistrosNativeInstaller.targets [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    <PropertyGroup>
+      <IsProducingAllLiuxDistrosNativeInstaller>true</IsProducingAllLiuxDistrosNativeInstaller>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="RunAllSandBoxAndPackage" >
+      <MSBuild Projects="$(MSBuildThisFileDirectory)/SandBoxAndPackageInParallel.proj"
```
[the remark this answers is not in the document]
Comment (wli3) [reply build/AllLinuxDistrosNativeInstaller.targets]: Moved to SandBoxAndPackageInParallel.proj, but there is not build speed improvement. I need more work on that, but I think it should not be an blocker for this PR

File: build/SandBoxAndPackageInParallel.proj (hunk 5 above)
Comment (wli3) [reply build/SandBoxAndPackageInParallel.proj]: Fixed

File: build/Package.targets [hunk 8]
```diff
@@ -7,10 +7,11 @@
   <Import Project="$(MSBuildThisFileDirectory)/package/Installer.MSI.targets" />
   <Import Project="$(MSBuildThisFileDirectory)/package/Installer.PKG.targets" />
   <Import Project="$(MSBuildThisFileDirectory)/package/Installer.RPM.targets" />
+  <Import Project="$(MSBuildThisFileDirectory)/package/CopySandBoxPackageOut.targets" />
 
   <Target Name="GenerateInstallers"
           Condition=" '$(SkipBuildingInstallers)' != 'true' "
-          DependsOnTargets="Init;Layout;GeneratePkgs;GenerateDebs;GenerateMsis;GenerateRpms" />
+          DependsOnTargets="Prepare;Layout;GeneratePkgs;GenerateDebs;GenerateMsis;GenerateRpms" />
```
Comment (dsplaisted) [inline build/Package.targets]: Why the change from Init to Prepare?

File: build/Publish.targets [hunk 9]
```diff
@@ -11,7 +11,6 @@
   <Target Name="Publish"
           Condition=" '$(PUBLISH_TO_AZURE_BLOB)' != '' "
           DependsOnTargets="Init;
-                            Package;
```
Comment (dsplaisted) [inline build/Publish.targets]: Why is Package removed here?

File: build/SandBoxAndPackage.proj [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  <PropertyGroup>
+    <SandBoxFolderName>bin/WorkingCopy</SandBoxFolderName>
+    <RelativeSandBoxPackageOutputFolder>$(SandBoxFolderName)/RelativeSandBoxPackageOutputFolder</RelativeSandBoxPackageOutputFolder>
+    <SandBoxRepoRoot>$(RepoRoot)/$(SandBoxFolderName)/$(LinuxDistrosNeedNativeInstaller)</SandBoxRepoRoot>
+    <NugetCacheFolderName>.nuget</NugetCacheFolderName>
+    <RelativeCompileOutputPath>bin/2</RelativeCompileOutputPath>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <_MirrorOfRepoRootForTools Include="$(RepoRoot)/**/*.*" Exclude="$(RepoRoot)/$(SandBoxFolderName)/**/*.*;$(RepoRoot)/nuget/**/*.*;$(RepoRoot)/bin/**/*.*" />
```
Comment (dsplaisted) [inline build/SandBoxAndPackage.proj]: Since the sandbox folder is under the bin folder, you can remove `$(RepoRoot)/$(SandBoxFolderName)/**/*.*` from the Exclude.

File: build/SandBoxAndPackage.proj [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    <_ArtifactsToPackage Include="$(RepoRoot)/$(RelativeCompileOutputPath)/$(Rid)/**/*.*"/>
+    <_ArtifactsToPackage Include="$(RepoRoot)/bin/Directory.Build.props"/>
```
Comment (dsplaisted) [inline build/SandBoxAndPackage.proj]: Does this mean that the Directory.Build.props file is included in the packages we generate?  Is that intentional?

File: build/SandBoxAndPackage.proj [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+    <Copy
+      SourceFiles="@(_NugetCacheToSpeedUp)"
+      DestinationFiles="@(_NugetCacheToSpeedUp -> '$(SandBoxRepoRoot)/$(NugetCacheFolderName)/%(RecursiveDir)%(Filename)%(Extension)')"/>
+
+  </Target>
+
+  <Target Name="PackageWithDocker" DependsOnTargets="PrepareSandBox">
+    <PropertyGroup>
+      <CommandToInvokeBuildScriptInDockerToPackageInSandBox>$(SandBoxRepoRoot)/build.sh --configuration $(BuildConfiguration) --docker $(DockerFolder) --skip-prereqs /target:GenerateInstallersAndCopyOutOfSandBox /p:RelativeSandBoxPackageOutputFolder=$(RelativeSandBoxPackageOutputFolder)</CommandToInvokeBuildScriptInDockerToPackageInSandBox>
```
Comment (dsplaisted) [inline build/SandBoxAndPackage.proj]: Nit: this is a super long line.  You could break it up into several lines by adding arguments, something like this:

```xml
<Command>build.sh --configuration $(BuildConfiguration)/<Command>
<Command>$(Command) -additionalParameter1 value1</Command>
<Command>$(Command) -additionalParameter2 value2</Command>
```

File: build/Package.targets (hunk 8 above)
Comment (wli3) [reply build/Package.targets]: "DownloadHostAndSharedFxArtifacts" is actually required for Package. And init does not have it

```xml
<Target Name="Prepare"
          DependsOnTargets="Init;DownloadHostAndSharedFxArtifacts;RestoreSrcPackages;RestoreToolsPackages;IsolateArtifactsFromDirectoryBuildFiles;CreatePropsForNextStage" />

  <Target Name="Init"
          DependsOnTargets="PrintBuildInfo;
                            SetTelemetryProfile;
                            BuildDotnetCliBuildFramework;
                            CheckPrereqs;">
```

File: build/Publish.targets (hunk 9 above)
Comment (wli3) [reply build/Publish.targets]: I swapped Package with "PackageAllLinux". So, if I do not remove it there, it will run the normal Package which I don't want, since I am reusing the package step.

I am not sure the best practice here. In a sense Publish does not require package (do nothing). The "normal" target, state explicit `compile, package, publish` so no behavior change

File: build/SandBoxAndPackage.proj (hunk 11 above)
Comment (wli3) [reply build/SandBoxAndPackage.proj]: Maybe a better name, it is for not looking for the props in upper directory. Not the actual subject of artifact

File: build/SandBoxAndPackage.proj (hunk 12 above)
Comment (wli3) [reply build/SandBoxAndPackage.proj]: Good trick!

File: build/AllLinuxDistrosNativeInstaller.targets [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Layout" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildThisFileDirectory)/LinuxDistrosNativeInstaller.props" />
+
+  <Target Name="AllLinuxDistrosNativeInstaller" DependsOnTargets="Prepare;Compile;RunAllSandBoxAndPackage;MultiPublish">
```
Comment (livarcocc) [inline build/AllLinuxDistrosNativeInstaller.targets]: Should this be BuildAndPublishAllLinuxDistrosNativeInstallers?

File: build/Prepare.targets [hunk 14]
```diff
@@ -13,6 +13,14 @@
                             CheckPrereqs;">
   </Target>
 
+  <Target Name="_SandBoxPrepare"
```
Comment (livarcocc) [inline build/Prepare.targets]: NIT: We don't release need the underscore on this target name.

File: build/SandBoxAndPackage.proj [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+      <CommandExitCodeFirstTime>0</CommandExitCodeFirstTime>
+      <PipeStderrToStdoutToCatchFirstFailure>2&gt;&amp;1</PipeStderrToStdoutToCatchFirstFailure>
+    </PropertyGroup>
+
+    <Exec Command="$(CommandToInvokeBuildScriptInDockerToPackageInSandBox) $(PipeStderrToStdoutToCatchFirstFailure)" ContinueOnError="WarnAndContinue">
```
Comment (livarcocc) [inline build/SandBoxAndPackage.proj]: We should find a way to communicate that this is a retry here. I was slightly confused. Maybe even a comment.

File: build/SandBoxAndPackage.proj [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+      <Output TaskParameter="ExitCode" PropertyName="CommandExitCodeFirstTime"/>
+    </Exec>
+
+    <Exec Command="$(CommandToInvokeBuildScriptInDockerToPackageInSandBox)" Condition=" '$(CommandExitCodeFirstTime)' != '0' "/>
+  </Target >
+
+  <Target Name="CopyPackageResult">
```
Comment (livarcocc) [inline build/SandBoxAndPackage.proj]: Do we need to make this conditional? What if the docker part failed?

File: scripts/obtain/dotnet-install.sh [hunk 17]
```diff
@@ -656,7 +656,9 @@ install_dotnet() {
     say_verbose "Zip path: $zip_path"
 
     say "Downloading link: $download_link"
-    download "$download_link" $zip_path || download_failed=true
+    # Failures are expected in the non-legacy case. Do not output to stderr, since
```
Comment (livarcocc) [inline scripts/obtain/dotnet-install.sh]: @johnbeisner make sure to review this part.

Comment (wli3) [reply scripts/obtain/dotnet-install.sh]: I discussed this line with him before

File: build/SandBoxAndPackage.proj (hunk 15 above)
Comment (wli3) [reply build/SandBoxAndPackage.proj]: Fixed by "extract property" https://github.com/dotnet/cli/pull/7615/commits/2ab9be0755a893ad77ccd9dd067830babe3ffba6

File: build/Prepare.targets (hunk 14 above)
Comment (wli3) [reply build/Prepare.targets]: Fixed, not _blah in this PR

File: build/AllLinuxDistrosNativeInstaller.targets (hunk 13 above)
Comment (wli3) [reply build/AllLinuxDistrosNativeInstaller.targets]: Fixed

File: build/SandBoxAndPackage.proj (hunk 12 above)
Comment (wli3) [reply build/SandBoxAndPackage.proj]: Done, fixed

File: build/SandBoxAndPackage.proj (hunk 11 above)
Comment (wli3) [reply build/SandBoxAndPackage.proj]: Fixed
